First-MVC series: Model & DB tutorials: Readability, MT, fix all issues (#21521)
dotnet/AspNetCore.Docs: aspnetcore/tutorials/first-mvc-app/adding-model.md

author: rick-anderson
description: Part 4 of tutorial series on ASP.NET Core MVC.
ms.author: riande
ms.date: 01/30/2021
no-loc: [appsettings.json, "ASP.NET Core Identity", cookie, Cookie, Blazor, "Blazor Server", "Blazor WebAssembly", "Identity", "Let's Encrypt", Razor, SignalR]
uid: tutorials/first-mvc-app/adding-model
---

# Part 4, add a model to an ASP.NET Core MVC app

In this section, classes are added for managing movies in a database. These classes are the "**M**odel" part of the **M**VC app.

These model classes are used with [Entity Framework Core](/ef/core) (EF Core) to work with a database. EF Core is an object-relational mapping (ORM) framework that simplifies the data access code that you have to write.

> The model classes you create are known as POCO classes (from **P**lain **O**ld **C**LR **O**bjects). POCO classes don't have any dependency on EF Core. They only define the properties of the data to be stored in the database.

Author: I thought about this one and purposely left this "you" in.  We have scaffolding that creates classes.  It could be clearer who is creating the classes.

author: rick-anderson
description: Part 4 of tutorial series on ASP.NET Core MVC.
ms.author: riande
ms.date: 02/01/2021
no-loc: [appsettings.json, "ASP.NET Core Identity", cookie, Cookie, Blazor, "Blazor Server", "Blazor WebAssembly", "Identity", "Let's Encrypt", Razor, SignalR]
uid: tutorials/first-mvc-app/adding-model
---

# Part 4, add a model to an ASP.NET Core MVC app

In this section, classes are added for managing movies in a database. These classes are the "**M**odel" part of the **M**VC app.

> These model classes are used with [Entity Framework Core](/ef/core) (EF Core) to work with a database. EF Core is an object-relational mapping (ORM) framework that simplifies the data access code that you have to write.

Author: @Rick-Anderson, I don't see a difference here with the text being suggested compared to the original.  Maybe there was something you meant to change but it didn't get in?

author: rick-anderson
description: Part 4 of tutorial series on ASP.NET Core MVC.
ms.author: riande
ms.date: 02/01/2021

> no-loc: [appsettings.json, "ASP.NET Core Identity", cookie, Cookie, Blazor, "Blazor Server", "Blazor WebAssembly", "Identity", "Let's Encrypt", Razor, SignalR]

Reviewer: Add Create, Edit, etc so they're not over translated.

---

# [Visual Studio](#tab/visual-studio)

Build the project as a check for compiler errors.

# [Visual Studio Code](#tab/visual-studio-code)

<a name="dc"></a>

## Create a database context class

> A database context class is needed to coordinate EF Core functionality (Create, Read, Update, Delete) for the `Movie` model. The database context is derived from [Microsoft.EntityFrameworkCore.DbContext](xref:Microsoft.EntityFrameworkCore.DbContext) and specifies the entities to include in the data model.

Reviewer: Yes, we want to use the scaffolder for everything. Open an issue for that on the RP series.

---

<a name="dc"></a>

> ## Create a database context class

Reviewer: replace below with examine - see lower comments for more details

A database context class is needed to coordinate EF Core functionality (Create, Read, Update, Delete) for the `Movie` model. The database context is derived from [Microsoft.EntityFrameworkCore.DbContext](/dotnet/api/microsoft.entityframeworkcore.dbcontext) and specifies the entities to include in the data model.

Create a *Data* folder.

Add a *Data/MvcMovieContext.cs* file with the following code: 

[!code-csharp[](~/tutorials/first-mvc-app/start-mvc/sample/MvcMovie3/zDocOnly/MvcMovieContext.cs?name=snippet)]

The preceding code creates a [DbSet\<Movie>](/dotnet/api/microsoft.entityframeworkcore.dbset-1) property for the entity set. In Entity Framework terminology, an entity set typically corresponds to a database table. An entity corresponds to a row in the table.

<a name="reg"></a>

> ## Register the database context

Reviewer: replace below with examine

Author: @Rick-Anderson, what are your referring to when you say "below".  Do you mean:
1.  Replace the next ## section _below_ this with "Examine"?  So "## Add a database connection string." would become "## Examine the database connection string".
2.  Replace the section _above_ "## Register the database context" so it is "Examine the database context".

I ask since it seemed like you might have meant "above" instead since this comment is right next to that, rather close to the next section below it.

Reviewer: Examine the db con str

Add the following `using` statements at the top of *Startup.cs*:

```csharp
using MvcMovie.Data;
using Microsoft.EntityFrameworkCore;
```

Add the following highlighted code in `Startup.ConfigureServices`:

# [Visual Studio](#tab/visual-studio)

[!code-csharp[](~/tutorials/first-mvc-app/start-mvc/sample/MvcMovie3/Startup.cs?name=snippet_ConfigureServices&highlight=5-6)]

# [Visual Studio Code / Visual Studio for Mac](#tab/visual-studio-code+visual-studio-mac)

[!code-csharp[](~/tutorials/first-mvc-app/start-mvc/sample/MvcMovie3/Startup.cs?name=snippet_UseSqlite&highlight=5-6)]

---

The name of the connection string is passed in to the context by calling a method on a [DbContextOptions](/dotnet/api/microsoft.entityframeworkcore.dbcontextoptions) object. For local development, the [ASP.NET Core configuration system](xref:fundamentals/configuration/index) reads the connection string from the *appsettings.json* file.

<a name="cs"></a>

> ## Add a database connection string

Reviewer: replace below with examine

> Visual Studio creates:

Reviewer: Move this below all the tabs and replace it with

Scaffolding creates:

* bla
* bla

Scaffolding updates

* bla
* bla

Open a command window in the project directory. The project directory is the directory that contains the *Program.cs*, *Startup.cs*, and *.csproj* files.

On Linux, export the scaffold tool path:

```console
export PATH=$HOME/.dotnet/tools:$PATH
```

Run the following command:

```dotnetcli
dotnet-aspnet-codegenerator controller -name MoviesController -m Movie -dc MvcMovieContext --relativeFolderPath Controllers --useDefaultLayout --referenceScriptLibraries
```

[!INCLUDE [explains scaffold generated params](~/includes/mvc-intro/model4.md)]

### [Visual Studio for Mac](#tab/visual-studio-mac)

Open a command window in the project directory. The project directory is the directory that contains the *Program.cs*, *Startup.cs*, and *.csproj* files.

Export the scaffold tool path:

```console
export PATH=$HOME/.dotnet/tools:$PATH
```

Run the following command:

```dotnetcli
dotnet-aspnet-codegenerator controller -name MoviesController -m Movie -dc MvcMovieContext --relativeFolderPath Controllers --useDefaultLayout --referenceScriptLibraries
```

[!INCLUDE [explains scaffold generated params](~/includes/mvc-intro/model4.md)]

---

<!-- End of tabs                  -->

> Scaffolding creates:

Reviewer: or updates

Reviewer: Update the appsettings the connection string. Some place you need to show that.
Update startup and show the changes someplace in this doc.

* A movies controller: *Controllers/MoviesController.cs*.
* Razor view files for Create, Delete, Details, Edit, and Index pages: `Views/Movies/\*.cshtml`.
* A database context class: *Data/MvcMovieContext.cs*.

The automatic creation of these files is known as *scaffolding*.

The scaffolded pages can't be used yet because the database doesn't exist. Running the app and selecting the **Movie App** link results in a *Cannot open database* or *no such table: Movie* error message.

<a name="dc"></a>

> ### Examine the generated database context class

Reviewer: ```suggestion
### Examine the generated database context class and reg
```
Move next section under this, too many `###`

A database context class is needed to coordinate EF Core functionality (Create, Read, Update, Delete) for the `Movie` model. The database context is derived from [Microsoft.EntityFrameworkCore.DbContext](xref:Microsoft.EntityFrameworkCore.DbContext) and specifies the entities to include in the data model.

> Scaffolding created the following database context class: Data/MvcMovieContext.cs:

Reviewer: Scaffolding creates the *Data/MvcMovieContext.cs* db ctx class:

[!code-csharp[](~/tutorials/first-mvc-app/start-mvc/sample/MvcMovie3/zDocOnly/MvcMovieContext.cs?name=snippet)]

> The preceding code creates a [DbSet\<Movie>](/dotnet/api/microsoft.entityframeworkcore.dbset-1) property for the entity set. In Entity Framework terminology, an entity set typically corresponds to a database table. An entity corresponds to a row in the table.

Reviewer: ```suggestion
The preceding code creates a [DbSet\<Movie>](/dotnet/api/microsoft.entityframeworkcore.dbset-1) property that represent collections of the specified entities in the db context.
```

<a name="reg"></a>

### Examine the generated database context registration

> ASP.NET Core is built with [dependency injection (DI)](xref:fundamentals/dependency-injection). Services, such as the EF Core DB context, must be registered with DI in `Startup`. Components that require these services are provided these services via constructor parameters. For example, Razor Pages frequently requires services from the DI container. The constructor code that gets a DB context instance is shown later in the tutorial. In this section, the database context is registered with the DI container.

Reviewer: Why are you talking about RP?

Reviewer: In the following code, the movie cntx is ...

Author: copy paste error, I forgot to remove the the mention of RP as the example.  Thanks for catching this.

Author: Could you be more explicit about what you to do with this: "In the following code, the movie cntx is ."  I'm sorry I am not catching on.  Sometimes brevity is not a friend to me here.     I can assume "cntx" means "context", but could you clarify what what you are suggesting here?

Reviewer: Replace

> The constructor code that gets a DB context instance is shown later in the tutorial. In this section, the database context is registered with the DI container.

with text code found later in the doc

In the Controllers/MoviesController.cs file, The constructor uses Dependency Injection to inject the MvcMovieContext database context into the controller. The database context is used in each of the CRUD methods in the controller.

Then remove the following section

# Dependency injection in the controller

Scaffolding added the following `using` statements at the top of *Startup.cs*:

```csharp
using MvcMovie.Data;
using Microsoft.EntityFrameworkCore;

> ```

Reviewer: Remove, too much of a distraction

Scaffolding generated the following highlighted code in `Startup.ConfigureServices`:

[!code-csharp[](~/tutorials/first-mvc-app/start-mvc/sample/MvcMovie3/Startup.cs?name=snippet_UseSqlite&highlight=5-6)]

> The name of the connection string is passed to the context by calling a method on a [DbContextOptions](/dotnet/api/microsoft.entityframeworkcore.dbcontextoptions) object. For local development, the [ASP.NET Core configuration system](xref:fundamentals/configuration/index) reads the `ConnectionString` key from the *appsettings.json* file.

Reviewer: The `[config system](xref:xyz0` reads the `"MvcMovieContext"` DB connection string. 

> For local development, the [ASP.NET Core configuration system](xref:fundamentals/configuration/index) reads the `ConnectionString` key from the *appsettings.json* file.

Move that to the next sections.

<a name="cs"></a>

### Examine the generated database connection string

Scaffolding added a connection string to the *appsettings.json* file:

> 

Reviewer: Need tabs here for both con strs

You can't use the scaffolded pages yet because the database doesn't exist. If you run the app and click on the **Movie App** link, you get a *Cannot open database* or *no such table: Movie* error message.
Scaffolding added a connection string to the *appsettings.json* file:

[!code-json[](~/tutorials/first-mvc-app/start-mvc/sample/MvcMovie3/appsettings_SQLite.json?highlight=10-12)]

<a name="migration"></a>

## Initial migration

Use the EF Core [Migrations](xref:data/ef-mvc/migrations) feature to create the database. Migrations is a set of tools that let you create and update a database to match your data model.
Use the EF Core [Migrations](xref:data/ef-mvc/migrations) feature to create the database. Migrations are a set of tools that create and update a database to match the data model.

# [Visual Studio](#tab/visual-studio)

From the **Tools** menu, select **NuGet Package Manager** > **Package Manager Console** (PMC).
From the **Tools** menu, select **NuGet Package Manager** > **Package Manager Console** .

In the PMC, enter the following commands:
In the Package Manager Console (PMC), enter the following command:

```powershell
Add-Migration InitialCreate

> Update-Database

Reviewer: I prefer them together. It flows better, they go together.

Reviewer: > I will put it back together but change the instruction to run them one at a time. 

No, just add an extra line with one space.

```
Update-Database
Update-Database
 
```

Results in

```
PM> Update-Database
Build started...
Build succeeded.
Done.
PM> Update-Database
Build started...
Build succeeded.
Done.
PM>  
``

Export the scaffold tool path:

```console
export PATH=$HOME/.dotnet/tools:$PATH
```

Run the following command:

```dotnetcli
dotnet-aspnet-codegenerator controller -name MoviesController -m Movie -dc MvcMovieContext --relativeFolderPath Controllers --useDefaultLayout --referenceScriptLibraries
```

[!INCLUDE [explains scaffold generated params](~/includes/mvc-intro/model4.md)]

<a name="sqlite-dev-vs-mac"></a>

### Use SQLite for development, SQL Server for production

When SQLite is selected, the template generated code is ready for development. The following code shows how to inject <xref:Microsoft.AspNetCore.Hosting.IWebHostEnvironment> into `Startup`. `IWebHostEnvironment` is injected so `ConfigureServices` can use SQLite in development and SQL Server in production.

[!code-csharp[](~/tutorials/first-mvc-app/start-mvc/sample/MvcMovie3/StartupDevProd.cs?name=snippet_StartupClass&highlight=3,5,10,16-28)]

---

<!-- End of tabs                  -->

<a name="scaffolding-created"></a>

Scaffolding creates:

* A movies controller: *Controllers/MoviesController.cs*.

> * Razor view files for Create, Delete, Details, Edit, and Index pages: `Views/Movies/\*.cshtml`.

Reviewer: remove `\`

Reviewer: ![image](https://user-images.githubusercontent.com/3605364/116488760-dfb3da00-a82e-11eb-8c36-3d132b211adb.png)

over translating. Helps to look at other languages

* A database context class: *Data/MvcMovieContext.cs*.

The automatic creation of these files is known as *scaffolding*.

The scaffolded pages can't be used yet because the database doesn't exist. Running the app and selecting the **Movie App** link results in a *Cannot open database* or *no such table: Movie* error message.

<a name="dc"></a>

### Examine the generated database context class and registration

> A database context class is needed to coordinate EF Core functionality (Create, Read, Update, Delete) for the `Movie` model. The database context is derived from [Microsoft.EntityFrameworkCore.DbContext](xref:Microsoft.EntityFrameworkCore.DbContext) and specifies the entities to include in the data model.

Reviewer: and delete

# [Visual Studio](#tab/visual-studio)

From the **Tools** menu, select **NuGet Package Manager** > **Package Manager Console** (PMC).

In the Package Manager Console (PMC), enter the following command:

```powershell
Add-Migration InitialCreate
```

* `Add-Migration InitialCreate`: Generates a *Migrations/{timestamp}_InitialCreate.cs* migration file. The `InitialCreate` argument is the migration name. Any name can be used, but by convention, a name is selected that describes the migration. Because this is the first migration, the generated class contains code to create the database schema. The database schema is based on the model specified in the `MvcMovieContext` class.

In the PMC, enter the following command:

```powershell
Update-Database
```

* `Update-Database`: Updates the database to the latest migration, which the previous command created. This command runs the `Up` method in the *Migrations/{time-stamp}_InitialCreate.cs* file, which creates the database.

The `Update-Database` command generates the following warning:

> > No type was specified for the decimal column 'Price' on entity type 'Movie'. This causes values to be silently truncated if they do not fit in the default precision and scale. Explicitly specify the SQL server column type that can accommodate all the values using 'HasColumnType()'.

Author: Thanks.  When checking for future tense I unfortunately forgot the warning explanation was also part of the actual output.

> [!NOTE]
> You may not be able to enter decimal commas in the `Price` field. To support [jQuery validation](https://jqueryvalidation.org/) for non-English locales that use a comma (",") for a decimal point and for non US-English date formats, the app must be globalized. For globalization instructions, see [this GitHub issue](https://github.com/dotnet/AspNetCore.Docs/issues/4076#issuecomment-326590420).

* Test the **Edit**, **Details**, and **Delete** pages.
Test the **Edit**, **Details**, and **Delete** pages.

## Dependency injection in the controller

# [Visual Studio](#tab/visual-studio)

Open the *Controllers/MoviesController.cs* file and examine the constructor:

<!-- l.. Make copy of Movies controller (or use the old one as I did in the 3.0 upgrade) because we comment out the initial index method and update it later  -->

[!code-csharp[](~/tutorials/first-mvc-app/start-mvc/sample/MvcMovie22/Controllers/MC1.cs?name=snippet_1)]

> The constructor uses [Dependency Injection](xref:fundamentals/dependency-injection) to inject the database context (`MvcMovieContext`) into the controller. The database context is used in each of the [CRUD](https://wikipedia.org/wiki/Create,_read,_update_and_delete) methods in the controller.

Reviewer: The original flows much better. Use bullets when there is a list of different things. In this case they are related. A primary reason for lists is to break up long sentences.

Reviewer: The constructor uses [Dependency Injection](xref:fundamentals/dependency-injection) to inject the `MvcMovieContext` database context into the controller. The database context is used in each of the [CRUD](https://wikipedia.org/wiki/Create,_read,_update_and_delete) methods in the controller.

> ### Use SQLite for development, SQL Server for production

Reviewer: This needs to be hidden from VS by a tab. It's too much of a distraction for SQl Server users.

All the manual stuff that was removed by scaffolding needs to be added back in, but this time, in lieu of 
# Create a database context class
change that to 
# Examine the generated database context class

Then under the new examine section on the DB context, you can add the SQL in production to the VSC/VSmac tabs.

Author: @Rick-Anderson,   

Putting together your comments around putting the scaffolding sections back in as "Examine" sections, I think I have that structured the way you want right now in the PR. 

However the instruction:  "Then under the new examine section on the DB context, you can add the SQL in production to the VSC/VSmac tabs." makes me worry I don't have it right.  What I ended up with would not have VSC/VSMac tabs at that point, unless I have the tabs in the wrong organization now.

Could you take a look and see if I have the structure right and you envisioned it, and  if so, specify exactly what line where you want the SQL in production section.  It is not clear to me, sorry to ask for clarification.  Handy review link: [Internal Review - Adding a Model](https://review.docs.microsoft.com/en-us/aspnet/core/tutorials/first-mvc-app/adding-model?view=aspnetcore-5.0&branch=pr-en-us-21521&tabs=visual-studio)

Reviewer: The scaffolded code is the same except for startup and the connection string - right? Use tabs where you need them.

Reviewer: > specify exactly what line where you want the SQL in production section. 

I took care of that in #21610

Author: Back to this.  I followed what you did in #21610.   Just a note though: I could not move the 4 line repeated text from the SQLite for dev... section to an include file I ended up repeating the 4 lines for VS Code and VS mac.   Oddly I if I removed those 4 lines, whether replacing them with an include fill or not, it would not merge.  I tested many times trying various ways to get it to allow a merge and it came down to removing text on lines 323 to 326 would result in a "Cannot merge commit" error.  I finally had to give in and just repeat the 4 lines between the two tabs, and worry about moving it to an include file (if that is absolutely necessary) after this PR.

* `InitialCreate.Up` creates the Movie table and configures `Id` as the primary key.
* `InitialCreate.Down` reverts the schema changes made by the `Up` migration.

<a name="test"></a>

## Test the app

Run the app and select the **Movie App** link.

If you get an exception similar to the following, you may have missed the [migrations step](#migration):

# [Visual Studio](#tab/visual-studio)

```console
SqlException: Cannot open database "MvcMovieContext-1" requested by the login. The login failed.
```

# [Visual Studio Code / Visual Studio for Mac](#tab/visual-studio-code+visual-studio-mac)

```console
SqliteException: SQLite Error 1: 'no such table: Movie'.
```

---

Test the **Create** page. Enter and submit data.

> [!NOTE]
> You may not be able to enter decimal commas in the `Price` field. To support [jQuery validation](https://jqueryvalidation.org/) for non-English locales that use a comma (",") for a decimal point and for non US-English date formats, the app must be globalized. For globalization instructions, see [this GitHub issue](https://github.com/dotnet/AspNetCore.Docs/issues/4076#issuecomment-326590420).

Test the **Edit**, **Details**, and **Delete** pages.

> ## Dependency injection in the controller

Reviewer: Move this up under Examine the generated database context registration where you're talking about it.

Examine the *Migrations/{timestamp}_InitialCreate.cs* migration file:

[!code-csharp[](~/tutorials/first-mvc-app/start-mvc/sample/MvcMovie5/Migrations/20201107014924_InitialCreate.cs?name=snippet)]

In the preceding code:

* `InitialCreate.Up` creates the Movie table and configures `Id` as the primary key.
* `InitialCreate.Down` reverts the schema changes made by the `Up` migration.

<a name="test"></a>

## Test the app

Run the app and select the **Movie App** link.

If you get an exception similar to one of the following, you may have missed the [migrations step](#migration):

# [Visual Studio](#tab/visual-studio)

```console
SqlException: Cannot open database "MvcMovieContext-1" requested by the login. The login failed.
```

# [Visual Studio Code / Visual Studio for Mac](#tab/visual-studio-code+visual-studio-mac)

```console
SqliteException: SQLite Error 1: 'no such table: Movie'.
```

---

Test the **Create** page. Enter and submit data.

> [!NOTE]

> > Decimal commas may not be allowed in the `Price` field. To support [jQuery validation](https://jqueryvalidation.org/) for non-English locales that use a comma (",") for a decimal point and for non US-English date formats, the app must be globalized. For globalization instructions, see [this GitHub issue](https://github.com/dotnet/AspNetCore.Docs/issues/4076#issuecomment-326590420).

Reviewer: Revert, original better

  ```console
  SqliteException: SQLite Error 1: 'no such table: Movie'.
  ```
```console
SqliteException: SQLite Error 1: 'no such table: Movie'.
```

---
  You probably missed the [migrations step](#migration).

* Test the **Create** page. Enter and submit data.

  > [!NOTE]
  > You may not be able to enter decimal commas in the `Price` field. To support [jQuery validation](https://jqueryvalidation.org/) for non-English locales that use a comma (",") for a decimal point and for non US-English date formats, the app must be globalized. For globalization instructions, see [this GitHub issue](https://github.com/dotnet/AspNetCore.Docs/issues/4076#issuecomment-326590420).

* Test the **Edit**, **Details**, and **Delete** pages.

## Dependency injection in the controller

# [Visual Studio](#tab/visual-studio)

Open the *Controllers/MoviesController.cs* file and examine the constructor:

<!-- l.. Make copy of Movies controller (or use the old one as I did in the 3.0 upgrade) because we comment out the initial index method and update it later  -->

[!code-csharp[](~/tutorials/first-mvc-app/start-mvc/sample/MvcMovie22/Controllers/MC1.cs?name=snippet_1)]

> The constructor uses [Dependency Injection](xref:fundamentals/dependency-injection) to inject the database context (`MvcMovieContext`) into the controller. The database context is used in each of the [CRUD](https://wikipedia.org/wiki/Create,_read,_update_and_delete) methods in the controller.

Reviewer: This needs to be restored. Why was it dropped?

Author: Thanks for catching this.  This is right where I ran into an odd cannot merge commit issue.  Just touching particular lines, no matter what I did would cause the error.  I ended up having to remove a section and build and test adding back to determine where it was hanging up.  I ended up accidently missing this section.

I should not have set this PR back from draft.  I thought it would only notify you for another review if I explicitly re-requested a review, sorry.  I was still testing through this set of instructions for all three IDE's/OS to make sure I didn't miss anything.

Earlier in this tutorial, you saw how a controller can pass data or objects to a view using the `ViewData` dictionary. The `ViewData` dictionary is a dynamic object that provides a convenient late-bound way to pass information to a view.

> MVC provides the ability to pass strongly typed model objects to a view. This strongly typed approach enables compile time code checking. The scaffolding mechanism used this approach. It passed a strongly typed model, with the `MoviesController` class and views.

Reviewer: ```suggestion
MVC provides the ability to pass strongly typed model objects to a view. This strongly typed approach enables compile time code checking. The scaffolding mechanism passed a strongly typed model in the `MoviesController` class and views.
```
per style guide, this is ambiguous

Examine the generated `Details` method in the *Controllers/MoviesController.cs* file:

[!code-csharp[](~/tutorials/first-mvc-app/start-mvc/sample/MvcMovie22/Controllers/MC1.cs?name=snippet_details)]

> The `id` parameter is generally passed as route data. For example `https://localhost:5001/movies/details/1` sets:

Reviewer: ```suggestion
The `id` parameter is generally passed as route data. For example, `https://localhost:5001/movies/details/1` sets:
```
Usually acrolinx suggests this.

* The controller to the `movies` controller, the first URL segment.
* The action to `details`, the second URL segment.
* The id to 1, the last URL segment.

The `id` can be passed in with a query string, as in the following example:

`https://localhost:5001/movies/details?id=1`

> The `id` parameter is defined as a [nullable type](/dotnet/csharp/programming-guide/nullable-types/index),`int?`. In case an id value isn't provided.

Reviewer: ```suggestion
The `id` parameter is defined as a [nullable type](/dotnet/csharp/programming-guide/nullable-types/index),`int?` to accommodate when the `id` value isn't provided.
```
**OR**
```suggestion
The `id` parameter is defined as a [nullable type](/dotnet/csharp/programming-guide/nullable-types/index),`int?` in cases when the `id` value isn't provided.
```

> A [lambda expression](/dotnet/articles/csharp/programming-guide/statements-expressions-operators/lambda-expressions) is passed in to `FirstOrDefaultAsync` to select movie entities that match the route data or query string value.

Reviewer: ```suggestion
A [lambda expression](/dotnet/articles/csharp/programming-guide/statements-expressions-operators/lambda-expressions) is passed in to <xref:System.Data.Entity.QueryableExtensions.FirstOrDefaultAsync%2A>to select movie entities that match the route data or query string value.
```

[!code-cshtml[](~/tutorials/first-mvc-app/start-mvc/sample/MvcMovie22/Views/Movies/IndexOriginal.cshtml?range=1)]

The `@model` directive allows access to the list of movies that the controller passed to the view by using a `Model` object that's strongly typed. For example, in the *Index.cshtml* view, the code loops through the movies with a `foreach` statement over the strongly typed `Model` object:

[!code-cshtml[](~/tutorials/first-mvc-app/start-mvc/sample/MvcMovie22/Views/Movies/IndexOriginal.cshtml?highlight=1,31,34,37,40,43,46-48)]

> Because the `Model` object is strongly typed as an `IEnumerable<Movie>` object, each item in the loop is typed as `Movie`. Among other benefits, the code gets compile time checking.

Reviewer: ```suggestion
Because the `Model` object is strongly typed as an `IEnumerable<Movie>` object, each item in the loop is typed as `Movie`. Among other benefits, the compiler validates the types used in the code.
```

[!code-cshtml[](~/tutorials/first-mvc-app/start-mvc/sample/MvcMovie22/Views/Movies/IndexOriginal.cshtml?range=1)]

The `@model` directive allows access to the list of movies that the controller passed to the view by using a `Model` object that's strongly typed. For example, in the *Index.cshtml* view, the code loops through the movies with a `foreach` statement over the strongly typed `Model` object:

[!code-cshtml[](~/tutorials/first-mvc-app/start-mvc/sample/MvcMovie22/Views/Movies/IndexOriginal.cshtml?highlight=1,31,34,37,40,43,46-48)]

> Because the `Model` object is strongly typed, as an `IEnumerable<Movie>` object, each item in the loop is typed as `Movie`. Among other benefits, this means that you get compile time checking of the code.

Reviewer: ```suggestion
Because the `Model` object is strongly typed as an `IEnumerable<Movie>` object, each item in the loop is typed as `Movie`. Among other benefits, the code gets compile time checking.
```
